include/nanobind/stl/bind_map.h

```diff
@@ -0,0 +1,196 @@
+/*
+    nanobind/stl/bind_map.h: nb::bind_map()
+
+    This implementation is a port from pybind11 with minimal adjustments.
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include <nanobind/nanobind.h>
+#include <nanobind/make_iterator.h>
+
+NAMESPACE_BEGIN(NB_NAMESPACE)
+NAMESPACE_BEGIN(detail)
+
+template <typename Map, typename Class_>
+void map_assignment(Class_ &cl)
+{
+    using KeyType = typename Map::key_type;
+    using MappedType = typename Map::mapped_type;
+    // Map assignment when copy-assignable: just copy the value
+    if constexpr(std::is_copy_assignable_v<MappedType>){
+        cl.def("__setitem__", [](Map &m, const KeyType &k, const MappedType &v)
```

Comment:
You can merge this `cl.def` with the one below and put the `if constexpr` *inside* the lanbda function.  The entire body of `map_assignment` can be merged into `bind_map`, there is no longer a reason to let this be a separate function given the `if constexpr` feature.

Reply:
More specifically: run `cl.def` *if* `std::is_copy_assignable_v<MappedType>` or `std::is_copy_constructible_v<MappedType>` holds. Then, choose the right way to do the assignment inside the lambda function with another `if constexpr`.

Reply:
Done in 312a763.